alphazero/_utils.py

```diff
@@ -26,3 +31,117 @@ def AlphaGoZeroLoss(
     q_value = output[1]
 
     return mse(policy_true, policy) + ratio * ce(q_true, q_value)
+
+
+class SelfPlayDataset(torch.utils.data.Dataset):
+    # untested
+    def __init__(self, game, root_dir, transform):
+        self.game = game
+        self.root_dir = root_dir
+        self.transform = transform
+
+    def __len__(self):
+        return sum(
+            [
+                len(os.listdir(directory))
+                for directory in os.listdir(self.root_dir)
+            ]
+        )
+
+    def __getitem__(self, idx):
+
+        if torch.is_tensor(idx):
+            idx = idx.tolist()
+
+        sample_loc = os.path.join(self.root_dir, idx[0])
+        sample = load_board(self.game, sample_loc)
+
+        if self.transform:
+            sample = self.transform(sample)
+
+        return sample
+
+
+def train(
+    game,
+    model,
+    optimizer,
+    n_games=1000,
+    subset_size=1000,
+    n_mcts_iter=1600,
+    temperature=1e-4,
+    device=torch.device("cpu"),
+    n_test_games=9,
+):
+
+    data = []
+    # generate data through self-play with MCTS
+    for i in range(n_games):
+        data.append(
+            self_play(
+                game, model, n_mcts_iter=n_mcts_iter, temperature=temperature
```

Comment:
One game will return multiple datapoints (one for every step, returns a list of datapoints in the form of tuples)

Reply:
Gotcha. Will you ping me once you have the API for self_play defined?